plugins/module_utils/template.py

```diff
@@ -0,0 +1,88 @@
+        self.summaries_path = "{0}/summaries".format(self.templates_path)
+        self.template = {}
+        self.template_path = ""
+        self.template_name = template_name
+        self.template_id = template_id
+
+        if template_id:
+            # Checking if the template with id exists to avoid error: MSO Error 400: Template ID 665da24b95400f375928f195 invalid
+            template_summary = self.mso.get_obj(self.summaries_path, templateId=self.template_id)
+            if template_summary:
+                self.template_path = "{0}/{1}".format(self.templates_path, self.template_id)
+                self.template = self.mso.query_obj(self.template_path)
+            else:
+                self.mso.fail_json(
+                    msg="Provided template id '{0}' does not exist. Existing templates: {1}".format(
+                        self.template_id,
+                        [
+                            "Template '{0}' with id '{1}'".format(template.get("templateName"), template.get("templateId"))
+                            for template in self.mso.query_objs(self.summaries_path)
+                        ],
+                    )
+                )
+        elif template_name:
+            if not template_type:
+                self.mso.fail_json(msg="Template type must be provided when using template name.")
+            template_summary = self.mso.get_obj(
+                self.summaries_path, templateName=self.template_name, templateType=TEMPLATE_TYPES[template_type]["template_type"]
+            )
+            if template_summary:
+                self.template_path = "{0}/{1}".format(self.templates_path, template_summary.get("templateId"))
+                self.template = self.mso.query_obj(self.template_path)
+        elif template_type:
+            self.template = self.mso.query_objs(self.summaries_path, templateType=TEMPLATE_TYPES[template_type]["template_type"])
+        else:
+            self.template = self.mso.query_objs(self.summaries_path)
+
+        # Remove unwanted keys from existing object for better output and diff compares
+        if isinstance(self.template, dict):
+            for key in ["_updateVersion", "version"]:
+                self.template.pop(key, None)
+
+    @staticmethod
+    def get_object_from_list(search_list, kv_list):
+        """
+        Get the first matched object from a list of mso object dictionaries.
+        :param search_list: Objects to search through -> List.
+        :param kv_list: Key/value pairs that should match in the object. -> List[KVPair(Str, Str)]
+        :return: The index and details of the object. -> Item (Named Tuple)
+                 Values of provided keys of all existing objects. -> List
+        """
+
+        def kv_match(kvs, item):
+            return all((item.get(kv.key) == kv.value for kv in kvs))
+
+        match = next((Item(index, item) for index, item in enumerate(search_list) if kv_match(kv_list, item)), None)
```

Comment:
I am wondering about the performance of this method. 
Would it be possible to have a very large `search_list` and `kv_list`?
If not then performance may not really be an issue.

Reply:
I can see from the other PR's that `kv_list` is mostly just 1 item like 'uuid' or 'name'.
But could `search_list` contain a large number of objects?

Reply:
There are occasions where it might be required, like we had on schema for schema id and template name for instance. It is most likely to be one item, but this provides the flexibility to search for more without adding new functionality. The problem is here that we need to match on values inside of lists instead of keys in dicts, due to how the response of schema/template APIs. If it is something you think is valuable we could have a look into performance analysis for this?

The search list size, depends on the configuration of customer. I do not know the exact size of this, but my assumption is that is the amount of objects will not even be thousands. Furthermore I think the bottle-neck with anything we do is not search operations but the I/O operations towards NDO.

Reply:
Where is this function used?

Reply:
If you refer to kv_match() that is used in line 76: `match = next((Item(index, item) for index, item in enumerate(search_list) if kv_match(kv_list, item)), None)`

Reply:
Sorry! I meant get_object_from_list(). I haven't checked the other PRs so maybe it is used there?

Reply:
Indeed see the following PRs:

https://github.com/CiscoDevNet/ansible-mso/pull/472/files#diff-3409e22b7188780872a6a0c90a47c28744a04759f0f5d53c5e5dfd5587fed0b4

https://github.com/CiscoDevNet/ansible-mso/pull/471/files#diff-3409e22b7188780872a6a0c90a47c28744a04759f0f5d53c5e5dfd5587fed0b4

https://github.com/CiscoDevNet/ansible-mso/pull/435/files#diff-3409e22b7188780872a6a0c90a47c28744a04759f0f5d53c5e5dfd5587fed0b4

https://github.com/CiscoDevNet/ansible-mso/pull/434/files#diff-3409e22b7188780872a6a0c90a47c28744a04759f0f5d53c5e5dfd5587fed0b4

Reply:
Makes sense, if search list is within the 100's I don't think we will see much of a performance impact with this code specifically. Maybe just something to take note of to look into if anyone does report performance issues.